src/DurableTask.Netherite/TransportLayer/EventHubs/PartitionProcessor.cs

```diff
@@ -293,142 +313,133 @@ async Task<PartitionIncarnation> StartPartitionAsync(PartitionIncarnation prior
-                }
-                else
-                {
-                    this.traceHelper.LogDebug("EventHubsProcessor {eventHubName}/{eventHubPartition} stopping partition (incarnation: {incarnation}, quickly: {quickly})", this.eventHubName, this.eventHubPartition, current.Incarnation, quickly);
-                    await current.Partition.StopAsync(quickly);
-                    this.traceHelper.LogDebug("EventHubsProcessor {eventHubName}/{eventHubPartition} stopped partition (incarnation {incarnation})", this.eventHubName, this.eventHubPartition, current.Incarnation);
-                }
+                    while (current != null && current.ErrorHandler.IsTerminated)
+                    {
+                        current = await current.Next;
+                    }
 
-                this.traceHelper.LogInformation("EventHubsProcessor {eventHubName}/{eventHubPartition} is shut down", this.eventHubName, this.eventHubPartition);
-            }
+                    if (current == null)
+                    {
+                        this.traceHelper.LogDebug("EventHubsProcessor {eventHubName}/{eventHubPartition} already canceled or terminated", this.eventHubName, this.eventHubPartition);
+                    }
+                    else
+                    {
+                        this.traceHelper.LogDebug("EventHubsProcessor {eventHubName}/{eventHubPartition} stopping partition (incarnation: {incarnation}, quickly: {quickly})", this.eventHubName, this.eventHubPartition, current.Incarnation, quickly);
+                        await current.Partition.StopAsync(quickly);
+                        this.traceHelper.LogDebug("EventHubsProcessor {eventHubName}/{eventHubPartition} stopped partition (incarnation {incarnation})", this.eventHubName, this.eventHubPartition, current.Incarnation);
+                    }
 
-            using (await this.deliveryLock.LockAsync())
-            {
-                if (this.shutdownTask == null)
-                {
-                    this.traceHelper.LogDebug("EventHubsProcessor {eventHubName}/{eventHubPartition} starting shutdown task", this.eventHubName, this.eventHubPartition);
-                    this.shutdownTask = Task.Run(() => ShutdownAsync());
-                }
-                else
-                {
-                    this.traceHelper.LogDebug("EventHubsProcessor {eventHubName}/{eventHubPartition} shutdown task already started", this.eventHubName, this.eventHubPartition);
+                    this.shutdownRegistration?.Dispose();
+
+                    // dispose the lock, after quickly acquiring and releasing it to make sure no one is still holding it
+                    (await this.deliveryLock.LockAsync()).Dispose();
+                    this.deliveryLock.Dispose();
+
+                    this.traceHelper.LogInformation("EventHubsProcessor {eventHubName}/{eventHubPartition} is shut down", this.eventHubName, this.eventHubPartition);
```

Comment:
this is a bit strange. Can you please clarify, in the code, why this is needed?

Reply:
Or ideally, can the code be refactored so we don't need this trick? I understand if not, it's just a bit odd

Reply:
It looks slightly wrong (double Dispose). Will fix that.
Also I will add a comment explaining it a bit more.